### DISCUSSION TITLE
Piggybank - Saved reset to 0

### DISCUSSION THREAD

**Discussion by bartekd123** [firefly-iii/firefly-iii#6657] [General]
Not sure if it is a bug, or I did something wrong. I see something and was wondering what the best way of resolving it. Bit of a weird issue, so starting as a discussion post.

Going into a piggy bank, I can see the following:
 - "Saved so Far" shows 0
 - Graph shows that it has $1278
 - Added transaction shows $639.40 (wrong, see next point)
 - Clicking on the Nov11th transaction brings me to the real transaction of $36.87 (Correct)

This is what I think caused it
 - I had Piggy banks setup for over a year with a target amount
 - Last week, a few of them I removed the target (setting target to 0)
 - Have an automation Rule that adds/removes the amount from a piggy bank based on criteria

Below are some screenshots of what I see:

Main Piggybank Dashboard:
<img width="1138" alt="image" src="https://user-images.githubusercontent.com/16650162/202882132-5bd0640a-fe4e-46c6-a816-a0e69ce307f7.png">

Individual Piggy Bank:
![image](https://user-images.githubusercontent.com/16650162/202883029-7a06adb8-7450-4c07-94e0-40ef083e7a36.png)

**Comment by JC5** [comment]
That could be a bug yes. The "saved so far" amount is leading. I think that if you set the target amount to zero it resets the amount but creates no (or a bad) event. I'll check it out!

**Reply by bartekd123** [reply]
Thanks. 

In that case, should I add the amount that was there before to resolve it, or should I wait for you to confirm just in case you want to see something with my data?
I have a few piggy banks that have the issue.

**Reply by JC5** [reply]
Yes, just add the amount that was there. I'll do some research myself, thanks for the offer!

**Reply by bartekd123** [reply]
It fixed the balance by adding the amount back in, but now the graph is messed up where it shows more that what is in there. Any way of fixing that?

**Reply by JC5** [reply]
Not from the UI, but you can remove entries from the "piggy_bank_events" table.

**Reply by bartekd123** [reply]
I deleted from that table. It deletes it from the list of transactions on the right hand pane.. but graph still shows the data. Is the graph generated from data in another table?

**Reply by JC5** [reply]
No, but it could be cached? Not really sure tho haha.

**Reply by bartekd123** [reply]
Hmm I think it is cached somewhere. The one I deleted the other day is clear now, but others I just deleted are still showing in the graph. I deleted a couple others in a different piggy bank, and it still shows in the graph, even in a different clean browsers. 

But all good, i think the others will clear up on their own. Thanks for your help on this one!

**Reply by bartekd123** [reply]
I just had my automated piggy bank transfer automation yesterday, and when I did my import that triggers my automation, it again reset them to 0, and shows an import of the balance of my piggy bank. 

Here is more details if you want to see what is happening.

<img width="879" alt="image" src="https://user-images.githubusercontent.com/16650162/204106554-219487b7-9061-40c3-aaa5-75f2680af288.png">

I just noticed something.. I still have a "target date" on the piggy bank. Could that cause this issue?

**Reply by bartekd123** [reply]
@JC5  any idea what I can do to fix my data? Or do I need to delete my piggy banks, and give up on the previous data?

**Reply by JC5** [reply]
If you can, try to collect some logs from around the time the automated piggy bank transfer automation fires again. I'm not sure what's happening but I'd like to find out.

**Reply by bartekd123** [reply]
I think this is what you are looking for in the log? The log is pretty long, even for a couple small transactions. Let me know if you are looking for a different section of the log


```
2022-12-12T20:11:30.806261474Z [2022-12-12 15:11:30] local.DEBUG: Executing rule action "update_piggy" with value "Misc Emergency"  
2022-12-12T20:11:30.806311471Z [2022-12-12 15:11:30] local.DEBUG: self::getActionClass("update_piggy") = "FireflyIII\TransactionRules\Actions\UpdatePiggybank"  
2022-12-12T20:11:30.806363855Z [2022-12-12 15:11:30] local.DEBUG: Triggered rule action UpdatePiggybank on journal #25281  
2022-12-12T20:11:30.806409490Z [2022-12-12 15:11:30] local.DEBUG: Found piggy bank #23 ("Misc Emergency")  
2022-12-12T20:11:30.806455144Z [2022-12-12 15:11:30] local.DEBUG: Piggy bank account is linked to source, so add amount.  
2022-12-12T20:11:30.806503123Z [2022-12-12 15:11:30] local.DEBUG: Max amount to add to piggy bank is 0.000000000000000000000000, amount is 21.010000000000  
2022-12-12T20:11:30.806551796Z [2022-12-12 15:11:30] local.DEBUG: Amount is now 0.000000000000000000000000  
2022-12-12T20:11:30.806598754Z [2022-12-12 15:11:30] local.WARNING: Amount left is zero, stop.  
2022-12-12T20:11:30.806642662Z [2022-12-12 15:11:30] local.DEBUG: Action "update_piggy" on journal #25281 was executed, so count a result. Updated transaction journal count is now 1.  
2022-12-12T20:11:30.806695032Z [2022-12-12 15:11:30] local.DEBUG: SearchRuleEngine:: done processing strict rule #138  
2022-12-12T20:11:30.806738856Z [2022-12-12 15:11:30] local.DEBUG: SearchRuleEngine:: rule #138 was triggered (on 1 transaction(s)).  
```

**Reply by JC5** [reply]
I've been working on a similar issue and I hope this will be fixed in the next release :+1:

**Reply by bartekd123** [reply]
Thanks!

**Comment by bartekd123** [comment]
I think this is a bug, not related to me resetting the target to 0. I ended up re-creating the piggy banks in my environment and the same thing happened again when I imported my transactions.

So I reproduced it in the demo environment to make sure it is nothing with my environment. This is what I did:

 - Created a piggy bank without a target amount
 - Added 10 euros to the piggy bank
 - Created a rule that matches a transfer and description to remove that amount from the piggy bank
 - Created a transfer

Actual behaviour: Piggy bank reset to 0
Expected Behaviour: It should add/subtract based on the transaction amount

Screenshots from demo environment:


<img width="1044" alt="image" src="https://user-images.githubusercontent.com/16650162/211169306-075d5058-2b79-422e-aa96-5395dc107019.png">


<img width="880" alt="image" src="https://user-images.githubusercontent.com/16650162/211169316-057e8e4a-7057-436f-b382-e147d918b08f.png">


<img width="392" alt="image" src="https://user-images.githubusercontent.com/16650162/211169327-13b20f5f-08b9-4dee-abfe-1fe02e3400d4.png">


<img width="1106" alt="image" src="https://user-images.githubusercontent.com/16650162/211169345-7d251ec8-3b87-4915-abd7-4f8536f27b3a.png">

**Reply by JC5** [reply]
Ah nice, I think I see what you mean :+1:. I believe the demo site handles this correctly now. Thanks for the excellent description!

**Comment by bartekd123** [comment]
Great! So you are able to fix it in the next release? I’ll test on the demo site later

**Reply by bartekd123** [reply]
tested in demo looks good there.

**Reply by JC5** [reply]
Great!
